### ISSUE TITLE
Source maps not working with Firefox

### ISSUE THREAD

**Issue by graue** [substack/node-browserify#681]
Firefox has partial support for source maps. The Console tab doesn't use them yet, but ClojureScript's source maps are working fine for me in the Debugger tab. Browserify's do not.

The issue seems to be full pathnames in the source map's "sources" key:

~~~.js
    "sources": [
        "/home/graue/code/projectname/node_modules/browserify/node_modules/browser-pack/_prelude.js",
        "/home/graue/code/projectname/node_modules/backbone/backbone.js",
        "/home/graue/code/projectname/node_modules/bean/bean.js",
        "/home/graue/code/projectname/node_modules/browserify/node_modules/buffer/index.js",
        "/home/graue/code/projectname/node_modules/browserify/node_modules/buffer/node_modules/base64-js/lib/b64.js",
        "/home/graue/code/projectname/node_modules/browserify/node_modules/buffer/node_modules/ieee754/index.js",
        ...
    ]
~~~

If I remove the beginnings of each path so all the pathnames are relative to the project root, e.g.

~~~.js
        "node_modules/browserify/node_modules/browser-pack/_prelude.js",
        "node_modules/backbone/backbone.js",
~~~

then base64-encode the modified source map and stick it back into the Browserify bundle, Firefox's debugger is happy (and Chrome's still is, too).

I've done this manually but may be able to work on a PR for this later, if making Browserify use relative paths is seen as an acceptable change and the best way to go here. Let me know if that sounds like a good plan.

**Comment by fitzgen** [comment]
I suspect this is https://bugzilla.mozilla.org/show_bug.cgi?id=1090768 / https://github.com/mozilla/source-map/pull/145 and if so should be fixed in 36+

If anyone can still reproduce in Nightly, please file a bug with the test case: https://bugzilla.mozilla.org/enter_bug.cgi?product=Firefox&component=Developer%20Tools%3A%20Debugger

**Comment by AbdullahDiaa** [comment]
Removing my home folder path from *[preludePath and row.sourceFile]* in `node_modules/gulp-browserify/node_modules/browserify/node_modules/browser-pack/index.js` solved the problem temporarily till the new update is pushed to firefox developer edition.
```js
sourcemap.addFile(
{ sourceFile: preludePath, source: prelude },
{ line: 0 }
);
...
sourcemap.addFile(
{ sourceFile: row.sourceFile, source: row.source },
{ line: lineno }
);
```

**Comment by terinjokes** [comment]
I'm not sure why Firefox would fetch the content when they're included already, and the linked bug suggests it's fixed in Firefox 37. A PR can probably be opened up against `browser-pack` to make the paths relative for older versions.

**Comment by jlongster** [comment]
I believe this is still a bug in our end of the Firefox debugger. Looks like we aren't correctly checking the inlined source content of the sourcemap. I will fix this. See https://bugzilla.mozilla.org/show_bug.cgi?id=1166048

Note that the source URLs included in a browserify sourcemap do have a bug though: a file URL is include like `/Users/james/foo/bar` but it should be `file:///Users/james/foo/bar`. That is a webpack bug, it needs to be a URL. The reason most people don't notice is because it also inlines the source content.

**Comment by jmm** [comment]
@jlongster 
> Note that the source URLs included in a browserify sourcemap do have a bug though: a file URL is include like `/Users/james/foo/bar` but it should be `file:///Users/james/foo/bar`.

What values are you referring to, the elements of `sources`?

>  That is a webpack bug, it needs to be a URL.

Do you mean browserify?

**Comment by zertosh** [comment]
@jlongster Ah yeah don't set a sourceRoot. That's an easy fix. We [already](https://github.com/substack/node-browserify/blob/b065829be44a214d2bb9817933aee988ff6d1599/index.js#L736) pass it to browser-pack, but don't do anything with it there. I'll test it later and merge a fix.

**Comment by jmm** [comment]
@zertosh 
> @jlongster Ah yeah don't set a sourceRoot. That's an easy fix. We already pass it to browser-pack, but don't do anything with it there. I'll test it later and merge a fix.

I'm not sure there's actually a bug here. I think the `sources` elements are always considered to be URLs (without protocol they're just relative to some base). And if the user agent never requests them, it doesn't matter to this degree what their values are. (A separate issue that does matter is if they leak information you don't want exposed, like absolute local paths).

**Comment by zertosh** [comment]
@jmm looking at the [source map spec](https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit), i think you're right:

> Line 4: An optional source root, useful for relocating source files on a server or removing repeated values in the “sources” entry.  This value is prepended to the individual entries in the “source” field.

so source root is optional.

> If the sources are not absolute URLs after prepending of the “sourceRoot”, the sources are resolved relative to the SourceMap (like resolving script src in a html document).

and the default behavior is sane. though we should use `sourceRoot` if provided in browser-pack, and remove it from the row.

**Comment by jmm** [comment]
@zertosh 
> though we should use sourceRoot if provided in browser-pack, and remove it from the row.

I haven't looked into it, but if browser-pack is supposed to accept a `sourceRoot` param and it's disregarding one passed to it, that does sound like a bug (and something that needs to be documented [you can see I currently have it stubbed in the [pipeline docs](http://jmm.github.io/browserify-pipeline-docs/)]). Seems odd for that to be a row prop though -- seems more appropriate as an opt. Is that what you meant by "remove it from the row"?

As long as browserify is generating source maps with `sourcesContent` I don't think it should automatically be populating `file://` in either `sourceRoot` or `sources`.

**Comment by zertosh** [comment]
> Seems odd for that to be a row prop though -- seems more appropriate as an opt. Is that what you meant by "remove it from the row"?

Yup and yup.

> As long as browserify is generating source maps with sourcesContent I don't think it should automatically be populating file:// in either sourceRoot or sources.

It should just be an option. I'll add it in browser-pack

**Comment by jmm** [comment]
I previously said:

> As long as browserify is generating source maps with `sourcesContent` I don't think it should automatically be populating `file://` in either `sourceRoot` or `sources`.

In case this comes up again in the future, I can't think of a reason for `file://` or *any* URI scheme to be included in the source map output by default, with or without `sourcesContent`. By default the output should be setup so that the scheme flows from the base URL used in the specification's resolution algorithm.

**Comment by jmm** [comment]
This issue has become too dated and muddled to be actionable at this point and there's at least one open bug on Firefox that may be the problem. Also, I don't know when it changed, but by default browserify currently generates source maps with relative `sources` values. For those reasons I'm closing this issue. To anyone who is still having a problem in Firefox >= 37:

If it seems to be a Firefox bug, please follow up there.

Otherwise, please open a new issue here with a minimal reproduction in a repo and any other relevant information, such as Firefox version.

See:
https://bugzilla.mozilla.org/show_bug.cgi?id=1166048
https://bugzilla.mozilla.org/show_bug.cgi?id=882986
https://bugzilla.mozilla.org/show_bug.cgi?id=852792
https://bugzilla.mozilla.org/show_bug.cgi?id=1090768

**Comment by Ghazgkull** [comment]
Looks like this problem was just fixed by Firefox in version 42.0a2. (Download the FirefoxDeveloperEdition)
